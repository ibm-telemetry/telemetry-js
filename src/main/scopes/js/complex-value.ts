/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
