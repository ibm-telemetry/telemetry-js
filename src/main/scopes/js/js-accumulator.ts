/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { JsImport } from '../js/interfaces.js'

/**
 * Base class for all JS Accumulators.
 * Responsible for maintaining an aggregated state of imports and other elements.
 */
export abstract class JsAccumulator {
  public imports: JsImport[]

  constructor() {
    this.imports = []
  }
}
