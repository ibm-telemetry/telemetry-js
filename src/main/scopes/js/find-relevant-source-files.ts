/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import getPropertyByPath = require('lodash/get.js')
import * as path from 'node:path'

import { ObjectPath } from 'object-scan'

import { getTrackedSourceFiles } from '../../core/get-tracked-source-files.js'
import { Logger } from '../../core/log/logger.js'
import { NoInstallationFoundError } from '../../exceptions/no-installation-found-error.js'
import { getDependencyTree } from '../npm/get-dependency-tree.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getInstalledVersionPaths } from '../npm/get-installed-version-paths.js'
import { getPackageData } from '../npm/get-package-data.js'
import { getPackageTrees } from '../npm/get-package-trees.js'
import { getTreePredecessor } from '../npm/get-tree-predecessor.js'
import { DependencyTree, PackageData } from '../npm/interfaces.js'

/**
 * Finds tracked source files and then filters them based on ones that appear in a  project which
 * depends on the in-context instrumented package/version.
 *
 * @param instrumentedPackage - Data about the instrumented package to use during filtering.
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param fileExtensions - List of file extensions to capture metrics for.
 * @param logger - Logger instance.
 * @returns A (possibly empty) array of source files.
 */
export async function findRelevantSourceFiles(
  instrumentedPackage: PackageData,
  cwd: string,
  root: string,
  fileExtensions: string[],
  logger: Logger
) {
  logger.traceEnter('', 'findRelevantSourceFiles', [instrumentedPackage, cwd, root, fileExtensions])
  const sourceFiles = await getTrackedSourceFiles(root, logger, fileExtensions)

  const dependencyTree = await getDependencyTree(cwd, root, logger)

  const filterPromises = sourceFiles.map(async (f) => {
    const prefix = await getDirectoryPrefix(path.dirname(f.fileName), logger)
    const prefixPackageData = await getPackageData(prefix, root, logger)

    let packageTrees = getPackageTrees(dependencyTree, prefixPackageData)

    let instrumentedInstallVersions: string[] | undefined = undefined
    let shortestPathLength: number | undefined = undefined
    do {
      for (const tree of packageTrees) {
        const instrumentedInstallPaths = getInstalledVersionPaths(tree, instrumentedPackage.name)
        if (instrumentedInstallPaths.length > 0) {
          const pathsLength = instrumentedInstallPaths[0]?.length ?? 0
          if (shortestPathLength === undefined || pathsLength < shortestPathLength) {
            instrumentedInstallVersions = instrumentedInstallPaths.map(
              (path) => getPropertyByPath(tree, path)['version']
            )
            shortestPathLength = pathsLength
          }
        }
      }
      // did not find, go up one level for all packages
      packageTrees = packageTrees
        .map((tree) => getTreePredecessor(dependencyTree, tree['path'] as ObjectPath))
        .filter((tree) => tree !== undefined) as DependencyTree[]
    } while (shortestPathLength === undefined && packageTrees.length > 0)

    if (instrumentedInstallVersions === undefined) {
      throw new NoInstallationFoundError(instrumentedPackage.name)
    }

    return instrumentedInstallVersions.some((version) => version === instrumentedPackage.version)
  })
  const filterData = await Promise.all(filterPromises)

  const results = sourceFiles.filter((_, index) => {
    return filterData[index]
  })

  logger.traceEnter('', 'findRelevantSourceFiles', results)
  return results
}
