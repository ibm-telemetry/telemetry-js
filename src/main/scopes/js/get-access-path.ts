/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import type { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { getNodeValueHandler } from './node-value-handler-map.js'

/**
 * Constructs a JsToken object from a given Identifier type AST node.
 *
 * @param node - TS node to retrieve access path for.
 * @param sourceFile - Root AST node.
 * @param logger - A logger instance.
 * @param currAccessPath - For internal use only, tracks current constructed access path.
 * @param topLevel - For internal use only, tracks top level function call.
 * @returns Access path to node represented as string array (chunks).
 */
export default function getAccessPath(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  logger: Logger,
  currAccessPath: string[] = [],
  topLevel: boolean = true
): string[] {
  switch (node.kind) {
    case ts.SyntaxKind.Identifier:
      return [...currAccessPath, (node as ts.Identifier).escapedText.toString()]
    case ts.SyntaxKind.PropertyAccessExpression:
      currAccessPath.push((node as ts.PropertyAccessExpression).name.escapedText.toString())
      break
    case ts.SyntaxKind.ElementAccessExpression: {
      const argumentExpression = (node as ts.ElementAccessExpression).argumentExpression
      const data = getNodeValueHandler(argumentExpression.kind, sourceFile, logger).getData(
        argumentExpression
      )
      if (data instanceof ComplexValue) {
        currAccessPath = [
          ...currAccessPath,
          getAccessPath(argumentExpression, sourceFile, logger, [], false).reverse().join('.')
        ]
      } else {
        currAccessPath.push(data?.toString() ?? '')
      }
      break
    }
  }

  let accessPath = currAccessPath

  if ('expression' in node) {
    accessPath = getAccessPath(
      node.expression as ts.Node,
      sourceFile,
      logger,
      currAccessPath,
      false
    )
  }

  return topLevel ? accessPath.reverse() : accessPath
}
