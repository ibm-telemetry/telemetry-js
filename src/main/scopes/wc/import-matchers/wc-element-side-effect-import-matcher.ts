/*
 * Copyright IBM Corp. 2023, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type { JsImport, JsImportMatcher } from '../../js/interfaces.js'
import { type WcElement } from '../interfaces.js'

/**
 * Finds a matching import for a WcElement from side-effect imports.
 *
 * @param element - The WcElement to find an import for.
 * @param imports - The list of JsImport objects collected from the file.
 * @returns A JsImport if matched, otherwise undefined.
 */
export class WcElementSideEffectImportMatcher implements JsImportMatcher<WcElement> {
  elementType: 'wc' = 'wc'
  /**
   * Determines if a given WcElement is a named import (e.g.: `import {something} from 'package'`).
   *
   * @param element - JsxElement to evaluate.
   * @param imports - Import elements to use for comparison.
   * @returns Corresponding JsImport if element was imported as a name import,
   * undefined otherwise.
   */
  findMatch(element: WcElement, imports: JsImport[]) {
    return imports.find(
      (i) =>
        !i.isDefault &&
        !i.isAll &&
        i.isSideEffect &&
        (i.name === element.name || `${i.prefix}-${i.name}` === element.name)
    )
  }
}
