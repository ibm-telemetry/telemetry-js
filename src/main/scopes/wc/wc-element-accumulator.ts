/*
 * Copyright IBM Corp. 2025, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { JsxElement } from 'typescript'

import type { JsImport } from '../js/interfaces.js'
import { JsAccumulator } from '../js/js-accumulator.js'
import type { CdnImport, WcElement } from './interfaces.js'

/**
 * Responsible for maintaining an aggregated state of imports and elements.
 */
export class WcElementAccumulator extends JsAccumulator {
  public cdnImports: CdnImport[]
  public readonly elements: (WcElement | JsxElement)[]
  public readonly elementImports: Map<WcElement | JsxElement, JsImport | CdnImport>
  public readonly scriptSources: string[]

  constructor() {
    super()
    this.cdnImports = []
    this.elements = []
    this.elementImports = new Map()
    this.scriptSources = []
  }
}
