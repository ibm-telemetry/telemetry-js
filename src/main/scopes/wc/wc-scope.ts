/*
 * Copyright IBM Corp. 2023, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { Trace } from '../../core/log/trace.js'
import { Scope } from '../../core/scope.js'
import { EmptyScopeError } from '../../exceptions/empty-scope.error.js'
import { findRelevantSourceFiles } from '../js/find-relevant-source-files.js'
import type { JsImportMatcher } from '../js/interfaces.js'
import { processFile } from '../js/process-file.js'
import { removeIrrelevantImports } from '../js/remove-irrelevant-imports.js'
import { JsxElementAllImportMatcher } from '../jsx/import-matchers/jsx-element-all-import-matcher.js'
import { JsxElementNamedImportMatcher } from '../jsx/import-matchers/jsx-element-named-import-matcher.js'
import { JsxElementRenamedImportMatcher } from '../jsx/import-matchers/jsx-element-renamed-import-matcher.js'
// import { type WcElement } from './interfaces.js'
import { type JsxElement } from '../jsx/interfaces.js'
import { ElementMetric } from '../jsx/metrics/element-metric.js'
import { getPackageData } from '../npm/get-package-data.js'
import type { PackageData } from '../npm/interfaces.js'
import { WcElementAccumulator } from './wc-element-accumulator.js'
import { jsxNodeHandlerMap } from './wc-node-handler-map.js'

/**
 * Scope class dedicated to data collection from a DOM-based environment.
 */
export class WcScope extends Scope {
  public static readonly fileExtensions = [
    '.js',
    '.mjs',
    '.cjs',
    '.jsx',
    '.tsx',
    '.mjsx',
    '.cjsx',
    '.mtsx',
    '.ctsx',
    '.html',
    '.htm'
  ]

  public override name = 'wc' as const
  private runSync = false

  /**
   * Entry point for the scope. All scopes run asynchronously.
   */
  @Trace()
  public override async run(): Promise<void> {
    const collectorKeys = this.config.collect[this.name]
    if (collectorKeys === undefined || Object.keys(collectorKeys).length === 0) {
      throw new EmptyScopeError(this.name)
    }

    const promises: Array<Promise<void>> = []

    Object.keys(collectorKeys).forEach((key) => {
      switch (key) {
        case 'elements':
          promises.push(this.captureAllMetrics())
          break
      }
    })

    await Promise.allSettled(promises)
  }

  /**
   * Generates metrics for all discovered instrumented jsx elements found in the current working
   * directory's project.
   */
  @Trace()
  async captureAllMetrics(): Promise<void> {
    const importMatchers = [
      new JsxElementAllImportMatcher(),
      new JsxElementNamedImportMatcher(),
      new JsxElementRenamedImportMatcher()
    ]
    const instrumentedPackage = await getPackageData(this.cwd, this.cwd, this.logger)
    const sourceFiles = await findRelevantSourceFiles(
      instrumentedPackage,
      this.cwd,
      this.root,
      WcScope.fileExtensions,
      this.logger
    )

    this.logger.debug('Filtered source files: ' + sourceFiles.map((f) => f.fileName))

    const promises: Promise<void>[] = []

    for (const sourceFile of sourceFiles) {
      // route JS/JSX differently

      if (sourceFile.fileName.endsWith('html')) {
      } else {
      }

      // if JS/JSX
      const resultPromise = this.captureFileMetrics(
        await sourceFile.createSourceFile(),
        instrumentedPackage,
        importMatchers
      )

      // if html
      //

      if (this.runSync) {
        await resultPromise
      } else {
        promises.push(resultPromise)
      }
    }

    await Promise.allSettled(promises)
  }

  /**
   * Generates metrics for all discovered instrumented jsx elements found
   * in the supplied SourceFile node.
   *
   * @param sourceFile - The sourcefile node to generate metrics for.
   * @param instrumentedPackage - Name and version of the instrumented package
   * to capture metrics for.
   * @param importMatchers - Matchers instances to use for import-element matching.
   */
  async captureFileMetrics(
    sourceFile: ts.SourceFile,
    instrumentedPackage: PackageData,
    importMatchers: JsImportMatcher<JsxElement>[]
  ) {
    const accumulator = new WcElementAccumulator()

    // TODO make a HTML node handler
    processFile(accumulator, sourceFile, jsxNodeHandlerMap, this.logger)

    this.logger.debug('Pre0filter accumulator contents:', JSON.stringify(accumulator))

    removeIrrelevantImports(accumulator, instrumentedPackage.name)
    this.resolveElementImports(accumulator, importMatchers)

    accumulator.elements.forEach((jsxElement) => {
      const jsImport = accumulator.elementImports.get(jsxElement)

      if (jsImport === undefined) {
        return
      }

      this.capture(
        new ElementMetric(jsxElement, jsImport, instrumentedPackage, this.config, this.logger)
      )
    })
  }

  resolveElementImports(
    accumulator: WcElementAccumulator,
    elementMatchers: JsImportMatcher<JsxElement>[]
  ) {
    accumulator.elements.forEach((jsxElement) => {
      const jsImport = elementMatchers
        .map((elementMatcher) => elementMatcher.findMatch(jsxElement, accumulator.imports))
        .find((jsImport) => jsImport !== undefined)

      if (jsImport === undefined) {
        return
      }

      accumulator.elementImports.set(jsxElement, jsImport)
    })
  }

  /**
   * **For testing purposes only.**
   * Makes the JsxScope collection run "synchronously" (one source file at a time). Defaults to
   * `false`.
   *
   * @param runSync - Boolean of whether or not to run synchronously.
   */
  setRunSync(runSync: boolean) {
    this.runSync = runSync
  }
}
