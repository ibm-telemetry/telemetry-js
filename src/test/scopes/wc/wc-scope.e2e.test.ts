/*
 * Copyright IBM Corp. 2025, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { type ConfigSchema } from '@ibm/telemetry-config-schema'
import { describe, expect, it } from 'vitest'

import { EmptyScopeError } from '../../../main/exceptions/empty-scope.error.js'
import type { JsImportMatcher } from '../../../main/scopes/js/interfaces.js'
import { JsxElementRenamedImportMatcher } from '../../../main/scopes/jsx/import-matchers/jsx-element-renamed-import-matcher.js'
import type { JsxElement } from '../../../main/scopes/jsx/interfaces.js'
import { WcElementSideEffectImportMatcher } from '../../../main/scopes/wc/import-matchers/wc-element-side-effect-import-matcher.js'
import type { WcElement } from '../../../main/scopes/wc/interfaces.js'
import { WcElementAccumulator } from '../../../main/scopes/wc/wc-element-accumulator.js'
import { WcScope } from '../../../main/scopes/wc/wc-scope.js'
import { clearDataPointTimes } from '../../__utils/clear-data-point-times.js'
import { clearTelemetrySdkVersion } from '../../__utils/clear-telemetry-sdk-version.js'
import { Fixture } from '../../__utils/fixture.js'
import { initLogger } from '../../__utils/init-logger.js'
import { initializeOtelForTest } from '../../__utils/initialize-otel-for-test.js'

const config: ConfigSchema = {
  projectId: 'abc123',
  version: 1,
  endpoint: '',
  collect: {
    wc: {
      elements: {
        allowedAttributeNames: ['firstProp', 'secondProp', 'firstprop', 'secondprop'],
        allowedAttributeStringValues: ['hi', 'wow']
      }
    }
  }
}

describe('class: WcScope', () => {
  const logger = initLogger()
  describe('run', () => {
    it('correctly captures wc element metric data', async () => {
      const metricReader = initializeOtelForTest().getMetricReader()
      const root = new Fixture('projects/web-components-project')
      const cwd = new Fixture('projects/web-components-project/node_modules/instrumented')
      const wcScope = new WcScope(cwd.path, root.path, config, logger)

      wcScope.setRunSync(true)
      await wcScope.run()

      const results = await metricReader.collect()

      clearTelemetrySdkVersion(results)
      clearDataPointTimes(results)

      expect(results).toMatchSnapshot()
    })

    it('throws EmptyScopeError if no collector has been defined', async () => {
      const fixture = new Fixture('projects/web-components-project/node_modules/instrumented')
      const root = new Fixture('projects/web-components-project')
      const scope = new WcScope(
        fixture.path,
        root.path,
        { collect: { npm: {} }, projectId: '123', version: 1, endpoint: '' },
        logger
      )

      scope.setRunSync(true)
      await expect(scope.run()).rejects.toThrow(EmptyScopeError)
    })

    it('only captures metrics for the instrumented package/version', async () => {
      let metricReader = initializeOtelForTest().getMetricReader()
      const root = new Fixture('projects/multiple-versions-of-instrumented-dep')
      const pkgA = new Fixture(
        'projects/multiple-versions-of-instrumented-dep/node_modules/instrumented'
      )
      const pkgB = new Fixture(
        'projects/multiple-versions-of-instrumented-dep/b/node_modules/instrumented'
      )

      let wcScope = new WcScope(pkgA.path, root.path, config, logger)
      wcScope.setRunSync(true)
      await wcScope.run()
      const resultsA = await metricReader.collect()

      metricReader = initializeOtelForTest().getMetricReader()

      wcScope = new WcScope(pkgB.path, root.path, config, logger)
      wcScope.setRunSync(true)
      await wcScope.run()
      const resultsB = await metricReader.collect()

      expect(resultsA.resourceMetrics.scopeMetrics[0]?.metrics[0]?.dataPoints).toHaveLength(1)
      expect(resultsB.resourceMetrics.scopeMetrics[0]?.metrics[0]?.dataPoints).toHaveLength(2)
    })

    it('captures metrics when instrumented package is installed in intermediate package', async () => {
      const metricReader = initializeOtelForTest().getMetricReader()
      const root = new Fixture('projects/hoisted-deeply-nested-deps')
      const cwd = new Fixture('projects/hoisted-deeply-nested-deps/node_modules/instrumented')
      const wcScope = new WcScope(cwd.path, root.path, config, logger)

      wcScope.setRunSync(true)
      await wcScope.run()

      const results = await metricReader.collect()

      clearTelemetrySdkVersion(results)
      clearDataPointTimes(results)

      expect(results).toMatchSnapshot()
    })

    it('captures metrics for workspace files when instrumented package is installed by root package', async () => {
      const metricReader = initializeOtelForTest().getMetricReader()
      const root = new Fixture('projects/workspace-files-governed-by-root-dep')
      const cwd = new Fixture(
        'projects/workspace-files-governed-by-root-dep/node_modules/instrumented-top-level'
      )
      const wcScope = new WcScope(cwd.path, root.path, config, logger)

      wcScope.setRunSync(true)
      await wcScope.run()

      const results = await metricReader.collect()

      clearTelemetrySdkVersion(results)
      clearDataPointTimes(results)

      expect(results).toMatchSnapshot()
    })
  })

  describe('resolveElementImports', () => {
    const wcScope = new WcScope('', '', config, logger)
    const namedImport = {
      name: 'name',
      path: 'instrumented',
      isDefault: false,
      isAll: false,
      isSideEffect: true
    }
    const renamedImport = {
      name: 'renameName',
      rename: 'rename',
      path: 'instrumented',
      isDefault: false,
      isAll: false,
      isSideEffect: true
    }
    const namedElement = {
      name: 'name',
      prefix: undefined,
      raw: '',
      attributes: []
    }
    const renamedElement = {
      name: 'rename',
      prefix: undefined,
      raw: '',
      attributes: []
    }

    it('correctly identifies elements with their matchers', () => {
      const accumulator = new WcElementAccumulator()
      accumulator.imports.push(namedImport)
      accumulator.imports.push(renamedImport)
      accumulator.elements.push(namedElement)
      accumulator.elements.push(renamedElement)

      wcScope.resolveElementImports(accumulator, [
        new WcElementSideEffectImportMatcher(),
        new JsxElementRenamedImportMatcher()
      ] as JsImportMatcher<JsxElement | WcElement>[])

      expect(accumulator.elementImports.get(namedElement)).toStrictEqual(namedImport)
      expect(accumulator.elementImports.get(renamedElement)).toStrictEqual(renamedImport)
    })

    it('discards elements that do not have a matcher', () => {
      const unmatchedElement1 = {
        name: 'noMatch1',
        prefix: undefined,
        raw: '',
        attributes: []
      }
      const unmatchedElement2 = {
        name: 'noMatch2',
        prefix: undefined,
        raw: '',
        attributes: []
      }
      const accumulator = new WcElementAccumulator()
      accumulator.imports.push(namedImport)
      accumulator.imports.push(renamedImport)
      accumulator.elements.push(namedElement)
      accumulator.elements.push(renamedElement)
      accumulator.elements.push(unmatchedElement1)
      accumulator.elements.push(unmatchedElement2)

      const wcScope = new WcScope('', '', config, logger)
      wcScope.resolveElementImports(accumulator, [new WcElementSideEffectImportMatcher()])
      expect(accumulator.elementImports.get(namedElement)).toStrictEqual(namedImport)
      expect(accumulator.elementImports.get(unmatchedElement1)).toBeUndefined()
      expect(accumulator.elementImports.get(unmatchedElement2)).toBeUndefined()
    })

    it('can accept empty array', () => {
      const accumulator = new WcElementAccumulator()
      expect(() => {
        wcScope.resolveElementImports(accumulator, [])
      }).not.toThrow()
    })
  })
})
