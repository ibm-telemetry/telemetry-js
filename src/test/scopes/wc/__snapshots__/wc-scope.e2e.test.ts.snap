// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: WcScope > run > captures metrics for workspace files when instrumented package is installed by root package 1`] = `
{
  "errors": [],
  "resourceMetrics": {
    "resource": Resource {
      "_asyncAttributesPromise": undefined,
      "_attributes": {
        "analyzed.commit": "commitHash",
        "analyzed.host": "host",
        "analyzed.owner": "owner",
        "analyzed.ownerPath": "host/owner",
        "analyzed.path": "host/owner/repository",
        "analyzed.refs": [],
        "analyzed.repository": "repository",
        "date": "1970-01-01T00:00:02.023Z",
        "environment.name": "test environment",
        "project.id": "projectId",
        "scan.id": "scan id",
        "service.name": "IBM Telemetry",
        "telemetry.emitter.name": "telemetryName",
        "telemetry.emitter.version": "telemetryVersion",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "[omitted for test snapshot]",
      },
      "_syncAttributes": {
        "analyzed.commit": "commitHash",
        "analyzed.host": "host",
        "analyzed.owner": "owner",
        "analyzed.ownerPath": "host/owner",
        "analyzed.path": "host/owner/repository",
        "analyzed.refs": [],
        "analyzed.repository": "repository",
        "date": "1970-01-01T00:00:02.023Z",
        "environment.name": "test environment",
        "project.id": "projectId",
        "scan.id": "scan id",
        "service.name": "IBM Telemetry",
        "telemetry.emitter.name": "telemetryName",
        "telemetry.emitter.version": "telemetryVersion",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "[omitted for test snapshot]",
      },
      "asyncAttributesPending": false,
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "npm.dependency.instrumented.name": "460da73f7e63aeb72106788158261156d165376633de24e4a6673122ce2470aa",
                  "npm.dependency.instrumented.raw": "460da73f7e63aeb72106788158261156d165376633de24e4a6673122ce2470aa",
                  "npm.dependency.instrumented.version.major": "1",
                  "npm.dependency.instrumented.version.minor": "0",
                  "npm.dependency.instrumented.version.patch": "0",
                  "npm.dependency.instrumented.version.raw": "92521fc3cbd964bdc9f584a991b89fddaa5754ed1cc96d6d42445338669c1305",
                  "wc.element.attributes.names": [
                    "firstProp",
                    "[redacted17]",
                    "[redacted19]",
                  ],
                  "wc.element.attributes.values": [
                    "[redacted16]",
                    "[redacted18]",
                    "[redacted6]",
                  ],
                  "wc.element.frameworkWrapper": "none",
                  "wc.element.module.specifier": "instrumented-top-level",
                  "wc.element.name": "PackageComp",
                },
                "endTime": [
                  0,
                  0,
                ],
                "startTime": [
                  0,
                  0,
                ],
                "value": 2,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "",
              "name": "wc.element",
              "type": "COUNTER",
              "unit": "",
              "valueType": 0,
            },
            "isMonotonic": true,
          },
        ],
        "scope": {
          "name": "wc",
          "schemaUrl": undefined,
          "version": "",
        },
      },
    ],
  },
}
`;

exports[`class: WcScope > run > captures metrics when instrumented package is installed in intermediate package 1`] = `
{
  "errors": [],
  "resourceMetrics": {
    "resource": Resource {
      "_asyncAttributesPromise": undefined,
      "_attributes": {
        "analyzed.commit": "commitHash",
        "analyzed.host": "host",
        "analyzed.owner": "owner",
        "analyzed.ownerPath": "host/owner",
        "analyzed.path": "host/owner/repository",
        "analyzed.refs": [],
        "analyzed.repository": "repository",
        "date": "1970-01-01T00:00:02.023Z",
        "environment.name": "test environment",
        "project.id": "projectId",
        "scan.id": "scan id",
        "service.name": "IBM Telemetry",
        "telemetry.emitter.name": "telemetryName",
        "telemetry.emitter.version": "telemetryVersion",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "[omitted for test snapshot]",
      },
      "_syncAttributes": {
        "analyzed.commit": "commitHash",
        "analyzed.host": "host",
        "analyzed.owner": "owner",
        "analyzed.ownerPath": "host/owner",
        "analyzed.path": "host/owner/repository",
        "analyzed.refs": [],
        "analyzed.repository": "repository",
        "date": "1970-01-01T00:00:02.023Z",
        "environment.name": "test environment",
        "project.id": "projectId",
        "scan.id": "scan id",
        "service.name": "IBM Telemetry",
        "telemetry.emitter.name": "telemetryName",
        "telemetry.emitter.version": "telemetryVersion",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "[omitted for test snapshot]",
      },
      "asyncAttributesPending": false,
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "npm.dependency.instrumented.name": "ec46e364d52dab2e207d33617267a84df5838926793c1b0a2974899fe28229f1",
                  "npm.dependency.instrumented.raw": "ec46e364d52dab2e207d33617267a84df5838926793c1b0a2974899fe28229f1",
                  "npm.dependency.instrumented.version.major": "1",
                  "npm.dependency.instrumented.version.minor": "0",
                  "npm.dependency.instrumented.version.patch": "0",
                  "npm.dependency.instrumented.version.raw": "92521fc3cbd964bdc9f584a991b89fddaa5754ed1cc96d6d42445338669c1305",
                  "wc.element.attributes.names": [
                    "firstProp",
                    "[redacted17]",
                    "[redacted19]",
                  ],
                  "wc.element.attributes.values": [
                    "[redacted16]",
                    "[redacted18]",
                    "[redacted6]",
                  ],
                  "wc.element.frameworkWrapper": "none",
                  "wc.element.module.specifier": "instrumented",
                  "wc.element.name": "PackageComp",
                },
                "endTime": [
                  0,
                  0,
                ],
                "startTime": [
                  0,
                  0,
                ],
                "value": 2,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "",
              "name": "wc.element",
              "type": "COUNTER",
              "unit": "",
              "valueType": 0,
            },
            "isMonotonic": true,
          },
        ],
        "scope": {
          "name": "wc",
          "schemaUrl": undefined,
          "version": "",
        },
      },
    ],
  },
}
`;

exports[`class: WcScope > run > correctly captures wc element metric data 1`] = `
{
  "errors": [],
  "resourceMetrics": {
    "resource": Resource {
      "_asyncAttributesPromise": undefined,
      "_attributes": {
        "analyzed.commit": "commitHash",
        "analyzed.host": "host",
        "analyzed.owner": "owner",
        "analyzed.ownerPath": "host/owner",
        "analyzed.path": "host/owner/repository",
        "analyzed.refs": [],
        "analyzed.repository": "repository",
        "date": "1970-01-01T00:00:02.023Z",
        "environment.name": "test environment",
        "project.id": "projectId",
        "scan.id": "scan id",
        "service.name": "IBM Telemetry",
        "telemetry.emitter.name": "telemetryName",
        "telemetry.emitter.version": "telemetryVersion",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "[omitted for test snapshot]",
      },
      "_syncAttributes": {
        "analyzed.commit": "commitHash",
        "analyzed.host": "host",
        "analyzed.owner": "owner",
        "analyzed.ownerPath": "host/owner",
        "analyzed.path": "host/owner/repository",
        "analyzed.refs": [],
        "analyzed.repository": "repository",
        "date": "1970-01-01T00:00:02.023Z",
        "environment.name": "test environment",
        "project.id": "projectId",
        "scan.id": "scan id",
        "service.name": "IBM Telemetry",
        "telemetry.emitter.name": "telemetryName",
        "telemetry.emitter.version": "telemetryVersion",
        "telemetry.sdk.language": "nodejs",
        "telemetry.sdk.name": "opentelemetry",
        "telemetry.sdk.version": "[omitted for test snapshot]",
      },
      "asyncAttributesPending": false,
    },
    "scopeMetrics": [
      {
        "metrics": [
          {
            "aggregationTemporality": 1,
            "dataPointType": 3,
            "dataPoints": [
              {
                "attributes": {
                  "npm.dependency.instrumented.name": "ec46e364d52dab2e207d33617267a84df5838926793c1b0a2974899fe28229f1",
                  "npm.dependency.instrumented.raw": "ec46e364d52dab2e207d33617267a84df5838926793c1b0a2974899fe28229f1",
                  "npm.dependency.instrumented.version.major": "0",
                  "npm.dependency.instrumented.version.minor": "1",
                  "npm.dependency.instrumented.version.patch": "0",
                  "npm.dependency.instrumented.version.raw": "6ad9613a455798d6d92e5f5f390ab4baa70596bc869ed6b17f5cdd2b28635f06",
                  "wc.element.attributes.names": [
                    "firstProp",
                    "secondProp",
                    "[redacted1]",
                    "[redacted3]",
                    "[redacted5]",
                    "[redacted7]",
                    "[redacted8]",
                    "[redacted10]",
                  ],
                  "wc.element.attributes.values": [
                    "hi",
                    "wow",
                    "[redacted2]",
                    "[redacted4]",
                    "[redacted6]",
                    "[redacted6]",
                    "[redacted9]",
                    "[redacted11]",
                  ],
                  "wc.element.frameworkWrapper": "none",
                  "wc.element.module.specifier": "instrumented",
                  "wc.element.name": "ImaginaryThing",
                },
                "endTime": [
                  0,
                  0,
                ],
                "startTime": [
                  0,
                  0,
                ],
                "value": 1,
              },
              {
                "attributes": {
                  "npm.dependency.instrumented.name": "ec46e364d52dab2e207d33617267a84df5838926793c1b0a2974899fe28229f1",
                  "npm.dependency.instrumented.raw": "ec46e364d52dab2e207d33617267a84df5838926793c1b0a2974899fe28229f1",
                  "npm.dependency.instrumented.version.major": "0",
                  "npm.dependency.instrumented.version.minor": "1",
                  "npm.dependency.instrumented.version.patch": "0",
                  "npm.dependency.instrumented.version.raw": "6ad9613a455798d6d92e5f5f390ab4baa70596bc869ed6b17f5cdd2b28635f06",
                  "wc.element.attributes.names": [
                    "firstprop",
                    "[redacted12]",
                    "secondprop",
                  ],
                  "wc.element.attributes.values": [
                    "hi",
                    "[redacted13]",
                    "wow",
                  ],
                  "wc.element.frameworkWrapper": "none",
                  "wc.element.module.specifier": "instrumented",
                  "wc.element.name": "accordion",
                },
                "endTime": [
                  0,
                  0,
                ],
                "startTime": [
                  0,
                  0,
                ],
                "value": 1,
              },
            ],
            "descriptor": {
              "advice": {},
              "description": "",
              "name": "wc.element",
              "type": "COUNTER",
              "unit": "",
              "valueType": 0,
            },
            "isMonotonic": true,
          },
        ],
        "scope": {
          "name": "wc",
          "schemaUrl": undefined,
          "version": "",
        },
      },
    ],
  },
}
`;
